src/ostrowski/rationals/basic.lean

```diff
@@ -0,0 +1,226 @@
+  abv_mul := λ x y, begin rw [trivial_abs], dsimp, split_ifs, any_goals {norm_num}, finish, finish, finish, finish, end }
+
+def padic_norm_ℝ (p: ℕ) := λ r: ℚ, (padic_norm p r: ℝ)
+
+instance abv_lift (abv: ℚ → ℚ) [habv: is_absolute_value abv]:
+  is_absolute_value (λ r: ℚ, (abv r: ℝ)) :=
+{ abv_nonneg := by { norm_cast, exact abv_nonneg abv, },
+  abv_eq_zero := by { norm_cast, intro x, exact abv_eq_zero abv, },
+  abv_add := by { norm_cast, exact abv_add abv, },
+  abv_mul := by { norm_cast, exact abv_mul abv, } }
+
+instance padic_is_abv {p: ℕ} [hp: nat.prime p] : is_absolute_value (padic_norm_ℝ p) :=
+  @abv_lift (padic_norm p) (@padic_norm.is_absolute_value p hp)
+
+lemma trivial_abs_is_one_iff_nonzero_arg (a: ℚ) : (a ≠ 0) ↔ (trivial_abs a = 1) :=
+begin
+    split,
+    intro h₁,
+    unfold trivial_abs,
+    by_cases (a = 0),
+    exfalso,
+    exact h₁ h,
+    rw if_neg h,
+    contrapose!,
+    intro h,
+    unfold trivial_abs,
+    rw if_pos h,
+    finish
+end
+
+lemma rat_mk_eq_div (x_num : ℤ) (x_denom : ℕ) (x_pos : 0 < x_denom) (x_cop : nat.coprime (int.nat_abs x_num) x_denom)
+: ({num := x_num, denom := x_denom, pos := x_pos, cop := x_cop}:ℚ) = (↑ x_num)/(↑ x_denom) := begin
+norm_cast,
+rw rat.mk,
+rw rat.mk_nat,
+rw dif_neg,
+rw rat.mk_pnat,
+{
+  simp [x_cop.gcd_eq_one],
+},
+exact ne_of_gt x_pos,
+end
+
+def hom_of_abv {α} [linear_ordered_field α] {β} [ring β] [nontrivial β]
+  (abv: β → α) [is_absolute_value abv]:
+  monoid_with_zero_hom β α := {
+    to_fun := abv,
+    map_zero' := abv_zero abv,
+    map_one' := abv_one abv,
+    map_mul' := abv_mul abv,
+ }
+
+def hom_of_equiv (abv: ℚ → ℝ) [habv: is_absolute_value abv] (α: ℝ) (α_ne_zero: α ≠ 0):
+  monoid_with_zero_hom ℚ ℝ := {
+    to_fun := (λ r: ℚ, (abv r) ^ α),
+    map_zero' := by {
+      rw abv_zero abv,
+      exact real.zero_rpow α_ne_zero,
+    },
+    map_one' := by {
+      rw abv_one abv,
+      exact real.one_rpow α,
+    },
+    map_mul' := by {
+      intros x y,
+      rw abv_mul abv,
+      rw real.mul_rpow (abv_nonneg abv x) (abv_nonneg abv y),
+    },
+  }
+
+-- Deserves its place in matlib, as `monoid_with_zero_hom.map_inv`
+theorem Γ₀_map_inv {G₁ G₂: Type} [group_with_zero G₁] [group_with_zero G₂]
+  (φ: monoid_with_zero_hom G₁ G₂) (a: G₁) (a_ne_zero: a ≠ 0): φ a⁻¹ = (φ a)⁻¹ :=
+begin
+  apply eq_inv_of_mul_left_eq_one,
+  rw ← monoid_with_zero_hom.map_mul,
+  rw inv_mul_cancel a_ne_zero,
+  rw monoid_with_zero_hom.map_one,
+end
+
+lemma induction_on_primes [P: nat → Prop] (h₁: P 0) (h₂: P 1)
+  (h: ∀ p a: ℕ, prime p → P a → P (p * a)): ∀ n: ℕ, P n :=
+begin
+  intro n,
+  apply unique_factorization_monoid.induction_on_prime,
+  exact h₁,
+  {
+    intros n h,
+    rw nat.is_unit_iff.1 h,
+    exact h₂,
+  },
+  {
+    intros a p _ hp ha,
+    exact h p a hp ha,
+  },
+end
+
+-- Integer values of a morphism `φ` and its value on `-1` completely determines `φ`
+theorem mul_mor_eq_of_eq_on_pnat (φ₁ φ₂: monoid_with_zero_hom ℚ ℝ)
```

Comment:
I'd also rename it into `ext_hom_pnat` or something like that as this is an extensional theorem in fact, right?